Pinta.Core/Effects/ColorGradient.cs

```diff
@@ -0,0 +1,183 @@
+			double newPosition = minPosition + adjustedOffset;
+			return KeyValuePair.Create (newPosition, stop.Color);
+		}
+
+		return new (
+			StartColor,
+			EndColor,
+			minPosition,
+			maxPosition,
+			sorted_stops.Select (ToNewStop)
+		);
+	}
+
+	public ColorBgra GetColor (double position)
+	{
+		if (position <= MinimumPosition) return StartColor;
+		if (position >= MaximumPosition) return EndColor;
+		if (sorted_stops.Count == 0) return HandleNoStops (position);
+		return HandleWithStops (position);
+	}
+
+	private ColorBgra HandleNoStops (double position)
+	{
+		double valueSpan = MaximumPosition - MinimumPosition;
+		double positionOffset = position - MinimumPosition;
+		double fraction = positionOffset / valueSpan;
+		return ColorBgra.Lerp (StartColor, EndColor, fraction);
+	}
+
+	private ColorBgra HandleWithStops (double position)
+	{
+		int immediateLowerIndex = BinarySearchLowerOrEqual (sorted_stops, position);
+		var immediatelyLower = immediateLowerIndex < 0 ? new GradientStop (MinimumPosition, StartColor) : sorted_stops[immediateLowerIndex];
+		if (immediatelyLower.Position == position) return immediatelyLower.Color;
+		int immediatelyHigherIndex = immediateLowerIndex + 1;
+		var immediatelyHigher = (immediatelyHigherIndex < sorted_stops.Count) ? sorted_stops[immediatelyHigherIndex] : new GradientStop (MaximumPosition, EndColor);
+		double valueSpan = immediatelyHigher.Position - immediatelyLower.Position;
+		double positionOffset = position - immediatelyLower.Position;
+		double fraction = positionOffset / valueSpan;
+		return ColorBgra.Lerp (immediatelyLower.Color, immediatelyHigher.Color, fraction);
+	}
+
+	private readonly record struct GradientStop (double Position, ColorBgra Color);
+
+	/// <returns>Index of number immediately lower than target. If not found, it returns -1</returns>
+	/// <param name="arr">Array assumed to be sorted by key</param>
+	private static int BinarySearchLowerOrEqual (IReadOnlyList<GradientStop> arr, double target)
+	{
+		// TODO: Make this method more generic?
+		if (arr.Count == 0) return -1;
+
+		int left = 0;
+		int right = arr.Count - 1;
+
+		while (left <= right) {
+			int mid = left + (right - left) / 2;
+			if (arr[mid].Position == target)
+				return mid;
+			else if (arr[mid].Position < target)
+				left = mid + 1;
+			else
+				right = mid - 1;
+		}
+
+		// If target is not found, 'left' will be the index of the number just greater than target
+		// so 'left - 1' will be the index of the number immediately lower than target
+		return left - 1;
+	}
+
+	private static ColorBgra DefaultStartColor () => ColorBgra.Black;
+	private static ColorBgra DefaultEndColor () => ColorBgra.White;
+	private static double DefaultMinimumValue () => 0;
+	private static double DefaultMaximumValue () => 255;
+	private static IEnumerable<KeyValuePair<double, ColorBgra>> EmptyStops () => Enumerable.Empty<KeyValuePair<double, ColorBgra>> ();
+
+	public static ColorGradient Create ()
```

Comment:
Is this empty Create() method useful for any specific purpose?

Reply:
It's just the default behavior if nothing is specified. In any case, I've just removed that one and other method that wasn't used, in order to keep the code slim.